### DISCUSSION TITLE
What happens if the primary database fails?

### DISCUSSION THREAD

**Discussion by benkingcode** [neondatabase/neon#9269] [Q&A]
I've been reading through lots of documentation trying to find an answer for this but there doesn't seem to be one. I understand that Neon is architected differently and as such doesn't have a traditional High Availability setup. But lets say I have a write and a read replica. What would happen if the write suddenly dies? Will it just be offline for a period until Neon spins up a new one? If so how long will that take?

I'm happy with a few seconds of read-only downtime, I'm just nervous about what exactly the worst case scenario is when it comes to the main/write database.

**Comment by manimovassagh** [comment]
Good question. Here is how Neon handles compute failures:

**What happens when the primary (write) compute goes down:**

1. **WAL is safe**: Your committed writes are stored durably in the safekeepers (Neon uses a quorum of 3 safekeepers, similar to Raft). Even if the compute dies mid-transaction, all committed data is safe.

2. **Automatic restart**: On the managed Neon service, the control plane detects the compute failure and spins up a new one. The new compute connects to the pageserver, which reconstructs pages from layer files in object storage plus recent WAL from safekeepers.

3. **Downtime duration**: Typically a few seconds to maybe 10-15 seconds for the new compute to become available, depending on the size of your working set. The compute is stateless, so it just needs to start Postgres and reconnect to the pageserver. Working set pages are fetched on demand.

4. **In-flight transactions are lost**: Any uncommitted transactions at the time of failure are rolled back, same as any Postgres crash recovery. But committed data is not lost.

**Read replicas during a primary failure:**

Read replicas in Neon connect to the same pageserver(s), not to the primary compute. So if the primary compute dies, read replicas can continue serving read queries for data up to the last WAL position they consumed. They will not get new data until the primary is back and producing WAL again.

**Worst case scenario:**

- A few seconds of total downtime (no reads or writes on primary) while the compute restarts
- Read replicas may continue serving slightly stale reads during this window
- No data loss for committed transactions
- Uncommitted transactions are rolled back

This is fundamentally better than traditional Postgres HA (Patroni/repmgr) where failover to a standby can take 10-30+ seconds and risks replication lag data loss. In Neon, there is no replication lag concern because the source of truth is the safekeepers plus object storage, not the compute itself.

The main risk factor is not the compute failing -- it is the pageserver or safekeeper quorum failing, which is a much rarer event on the managed service since those components are distributed and redundant.
